src/gt4py/next/common.py

```diff
@@ -117,10 +116,143 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:
+            continue
+        # create a vertex for each dimension
+        for dim in dims:
+            graph.setdefault(dim, set())
+        # add edges
+        predecessor = dims[0]
+        for dim in dims[1:]:
+            graph[dim].add(predecessor)
+            predecessor = dim
+
+    # modified version of Kahn's algorithm
+    topologically_sorted_list: list[Dimension] = []
+
+    # compute in-degree for each vertex
+    in_degree = {v: 0 for v in graph.keys()}
+    for v1 in graph:
+        for v2 in graph[v1]:
+            in_degree[v2] += 1
+
+    # process vertices with in-degree == 0
+    # TODO(tehrengruber): avoid recomputation of zero_in_degree_vertex_list
+    while zero_in_degree_vertex_list := [v for v, d in in_degree.items() if d == 0]:
+        if len(zero_in_degree_vertex_list) != 1:
+            raise ValueError(
+                f"Dimensions can not be promoted. Could not determine "
+                f"order of the following dimensions: "
+                f"{', '.join((dim.value for dim in zero_in_degree_vertex_list))}."
+            )
+        v = zero_in_degree_vertex_list[0]
+        del in_degree[v]
+        topologically_sorted_list.insert(0, v)
+        # update in-degree
+        for predecessor in graph[v]:
+            in_degree[predecessor] -= 1
+
+    if len(in_degree.items()) > 0:
+        raise ValueError(
+            f"Dimensions can not be promoted. The following dimensions "
+            f"appear in contradicting order: {', '.join((dim.value for dim in in_degree.keys()))}."
+        )
+
+    return topologically_sorted_list
+
+
+class Domain(Collection):
+    dims: tuple[Dimension]
+    ranges: tuple[UnitRange]
+
+    def __init__(self, dims: tuple[Dimension, ...], ranges: tuple[UnitRange, ...]):
+        self.dims = dims
+        self.ranges = ranges
+
+    def __len__(self) -> int:
+        return len(self.ranges)
+
+    def __iter__(self):
+        return iter(zip(self.dims, self.ranges))
+
+    def __contains__(self, item: int | UnitRange) -> bool:
+        if isinstance(item, int):
+            return any(item in range_ for range_ in self.ranges)
+        elif isinstance(item, UnitRange):
+            return any(
+                item.start >= range_.start and item.stop <= range_.stop for range_ in self.ranges
+            )
+        return False
+
+    def __and__(self, other: Domain) -> Domain:
+        broadcast_dims = tuple(promote_dims(self.dims, other.dims))
+        broadcasted_ranges_self = self.broadcast_ranges(broadcast_dims)
+        broadcasted_ranges_other = other.broadcast_ranges(broadcast_dims)
+
+        intersected_ranges = []
+        for range1, range2 in zip(broadcasted_ranges_self, broadcasted_ranges_other):
+            intersected_range = range1 & range2
+            intersected_ranges.append(intersected_range)
+
+        return Domain(broadcast_dims, tuple(intersected_ranges))
+
+    def broadcast_ranges(self, broadcast_dims: tuple[Dimension, ...]) -> tuple[UnitRange, ...]:
```

Comment:
```suggestion
    def broadcasted_ranges(self, broadcast_dims: tuple[Dimension, ...]) -> tuple[UnitRange, ...]:
```
or make a free function for clarity (as it's not changing the current instance.